design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,777 @@
+# also aggregates to the crossplane-edit role (and thus transitively to the
+# crossplane-admin role).
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:composite:examplecomposites.xr.example.org:aggregate-to-edit
+  labels:
+    rbac.crossplane.io/aggregate-to-edit: "true"
+    rbac.crossplane.io/aggregate-to-ns-edit: "true"
+    rbac.crossplane.io/xrd: examplecomposites.xr.example.org
+rules:
+- apiGroups: [xr.example.org]
+  resources:
+  - exampleclaims
+  - exampleclaims/status
+  # Composite resources are cluster scoped. Binding this cluster role at the
+  # namespace scope will not actually grant access to these composite resources.
+  # Including them in this cluster role allows one crossplane:composite role to
+  # aggregate to both the crossplane-ns-example-edit and crossplane-edit roles.
+  - examplecomposites
+  - examplecomposites/status
+  verbs: ["*"]
+```
+
+### Provider ClusterRole Mechanics
+
+When a provider is installed (via the package manager) the RBAC manager will
+define three cluster roles:
+
+* `crossplane:provider:7f63e3661:aggregate-to-edit`
+* `crossplane:provider:7f63e3661:aggregate-to-view`
+* `crossplane:provider:7f63e3661:system`
+
+Note that these role names are derived from the active `ProviderRevision`. In
+this example `7f63e3661` is an abbreviated hash. The `aggregate-to-edit` role
+aggregates to the `crossplane`, `crossplane-admin` and `crossplane-edit` roles,
+while the `aggregate-to-view` role aggregates to the `crossplane-view` role. The
+`system` role does not aggregate; it is granted by the RBAC manager to the
+service account under which the provider pod runs.
+
+To create a cluster role for a `Provider` (package), the RBAC manager:
+
+1. Finds its active `ProviderRevision`.
+1. Lists all CRDs that are controlled by that `ProviderRevision`.
+1. Creates a `ClusterRole` that grants access to the defined CRs.
+
+In the case of the `system` role the RBAC manager creates a `ClusterRoleBinding`
+to the `ServiceAccount` under which the provider runs. This service account is
+also controlled by the aforementioned `ProviderRevision`, and can be discovered
+similarly.
+
+Below is an example of the `system` provider role and its role binding.
+
+```yaml
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crossplane:provider:7f63e3661:system
+rules:
+- apiGroups: [provider.example.org]
+  resources:
+  - examplemanageds
+  - examplemanageds/status
+  - exampleproviderconfigs
+  - exampleproviderconfigs/status
+  verbs: [get, list, watch, update, patch]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: crossplane:provider:7f63e3661:system
+roleRef:
+  apiVersion: rbac.authorization.k8s.io/v1
+  kind: ClusterRole
+  name: crossplane:provider:7f63e3661:system
+subjects:
+- kind: ServiceAccount
+  name: crossplane:provider:7f63e3661
```

Comment:
Any preference on whether I name this just `7f63e3661` or `crossplane-provider-7f63e3661`? (Or something else?)

Reply:
I am currently just naming it the hash, but it may be confusing if someone is looking at all of the SA's present in their cluster to just see a bunch of long hash names. I like `crossplane-provider-*`. If we adjust to a different name we can always circle back and update here :+1: